akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/StashingBehaviors.scala

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2017-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.scaladsl
+
+import java.util.concurrent.TimeoutException
+
+import akka.actor.typed.{ ActorSystem, Behavior }
+import akka.util.{ LineNumbers, PrettyDuration }
+
+import scala.concurrent.{ ExecutionContext, ExecutionContextExecutor, Future }
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * Stashing convenience behaviors that encapsulate common stashing patterns like awaiting a future to complete,
+ * or collecting a number of futures or responses before continuing with further handling of incoming messages.
+ */
+object StashingBehaviors {
+
+  private sealed trait Protocol
+  private final case class FutureCompleted[T](value: Try[T]) extends Protocol
+
+  final class StashFutureFailedException(msg: String, cause: Throwable)
+    extends RuntimeException(msg, cause)
+
+  /**
+   * Keep stashing incoming messages until the passed in [[scala.concurrent.Future]] `f` completes successfully.
+   *
+   * See also [[untilSuccessful]] if you want to react only to the the future's success, and treat failure by
+   * escalating it by failing the actor.
+   */
+  def until[T, M](f: Future[T], buffer: StashBuffer[M], timeout: FiniteDuration)(next: Try[T] ⇒ Behavior[M]): Behavior[M] = {
+    Behaviors.setup[M] { context ⇒
+      def name: String = s"until($f, $buffer, $timeout)(${LineNumbers(next)})"
+      implicit def ec: ExecutionContextExecutor = context.executionContext
+
+      f.value match {
+        case Some(completed) ⇒ next(completed) // fast-path for already completed future
+        case _ ⇒
+          pipeToSelf(f, timeout, context, name _)
+          Behaviors.receiveMessage[Any] {
+            case FutureCompleted(t: Try[T @unchecked]) ⇒
+              buffer.unstashAll(context, next(t)).asInstanceOf[Behavior[Any]]
+            case msg: M @unchecked ⇒
+              buffer.stash(msg)
+              Behaviors.same
+          }.narrow[M]
+      }
+    }
+  }
+
+  // TODO implement one in terms of the other
+
+  // TODO implement collect(f1, f2, f3, (f1, f2, f3) => Behavior)
```

Comment:
User could just do that themselves with future.zip:ing a bunch of futures etc. I don't think such future composition belongs in this API.

Reply:
Telling people to zip futures does not work nicely for more than 2 elements:

```
    val a = Future(":")
    val b = Future(12)
    val c = Future(12)
    
    val eventualTuple1: Future[(String, Int)] = a.zip(b)
    val eventualTuple2: Future[((String, Int), Int)] = a.zip(b).zip(c) // meh
```

which is why I proposed separate API for it; though I'm ok with waiting until people actually demand this.

Future.sequence does not work nicely either, since different types of `a` `b` `c`

Reply:
Then weed need to do that **anywhere** we accept a future. I don't think that is a good idea.

Btw: https://github.com/johanandren/futiles#multiple-futures-into-one-future-tuple

Reply:
I'd say it would not imply everywhere; here I'd see it as more common since it is the typical "gather some stuff" where the stuff is very likely completely different types. Only other style of such thing in actors I suppose is persistAll, but there all events are of the same type E